Feat/issue 121/extract vertical lines

letao opened the pull request:
Addresses https://github.com/swisstopo/swissgeol-boreholes-dataextraction/issues/121

Extracts and plots vertical lines as part of the extract and plot the vertical lines as part of the extract_lines() function. The line extracted are currently not available as part of the process_page() function and only serves as illustrations in the artifacts. In future issues the lines can be passed to process_page() function where corresponding methods for their usage need to be developed:

https://github.com/swisstopo/swissgeol-boreholes-dataextraction/blob/bb4e64f9115ada2c3cc4a64de9e0742bdeabd656/src/extraction/main.py#L274-L278

To keep the compatibility with current approach the naming of the non_vertical lines is kept as geometric_lines in the main.py script

The lines are drawn automatically when the extraction pipeline is executed with the "-l" flag.

stijnvermeeren-swisstopo commented on src/extraction/features/utils/geometry/linesquadtree.py:
```diff
@@ -101,7 +101,7 @@ def _qtree_insert(self, point: Point, line_key: str):
         coordinates = (round(point.x), round(point.y))
         qtree_point = self.qtree.find(coordinates)
-        if qtree_point:
+        if qtree_point in qtree_point.data:
```
How can an object be contained in its own attribute?

letao replied:
This was an oversight from my part, the line is missing a crucial element that was accidentally deleted:

if qtree_point and line_key in qtree_point.data:

The aim of this was to address a key error that appears when dealing with both vertical and non_vertical lines in the pipeline call. I believe there is a a chance that the same data gets processed more than once resulting in the key error.

stijnvermeeren-swisstopo commented on src/extraction/features/utils/geometry/geometric_line_utilities.py:
```diff
@@ -14,20 +14,38 @@
+    return non_vertical_lines, vertical_lines
+
+
 def drop_vertical_lines(lines: list[Line], threshold: float = 0.1) -> ArrayLike:
```
method not used anymore (only in unit test)?

stijnvermeeren-swisstopo commented on src/extraction/annotations/plot_utils.py:
```diff
@@ -76,21 +76,44 @@ def _convert_line_to_grid(line: Line, scale_factor: float) -> Line:
     return open_cv_img
 
 
+def plot_single_line_set(page: pymupdf.Page, lines: list[Line], scale_factor: float = 2.0) -> np.ndarray:
```
method not used?

letao replied:
This has been removed

stijnvermeeren-swisstopo reviewed (changes requested):
Is there a reason for keeping the vertical and non-vertical lines in separate variables throughout the code? It seems to me that the code would be simpler if you just have one list `geometric_lines` with both vertical and non-vertical lines. Then you could for examples add a property `is_vertical` on the `Line` object, so that you can easily filter out the vertical lines in those places where we explicitly don't want to consider them. Wouldn't that be simpler, or am I missing something?

letao commented:
The aim of the current approach of having separate variable throughout the code was to avoid any interaction with current functionality. Vertical lines were fully isolated from non_vertical. This might have been done in a way that is not too elegant.

I have proposed an alternative approach where the line separation is addressed within the line_detection.py script using the line_type parameter:

https://github.com/swisstopo/swissgeol-boreholes-dataextraction/blob/bd94c7d5f5dd69d24fe4fc9e40f14913b404546a/src/extraction/features/utils/geometry/line_detection.py#L75-L102

This allows us to select the specific types of lines we want to extract in the main script:

https://github.com/swisstopo/swissgeol-boreholes-dataextraction/blob/bd94c7d5f5dd69d24fe4fc9e40f14913b404546a/src/extraction/main.py#L273-L278

While also allowing us to select additional lines to plot in the artifact:

https://github.com/swisstopo/swissgeol-boreholes-dataextraction/blob/c41a60eafbbb37aa3bc75b26c6311dc02817848d/src/extraction/main.py#L304-L310

In my opinion this keeps the functionality while also allowing us to quickly choose the way we separate the lines. If it is still preferable to have a separate property in the Line object it can be changed.

letao commented:
@stijnvermeeren-swisstopo 

Two new issues addressing the comments have been created:

https://github.com/swisstopo/swissgeol-boreholes-dataextraction/issues/206 and https://github.com/swisstopo/swissgeol-boreholes-dataextraction/issues/207.

I was not able to reproduce the issue with 691251009-bp.pdf, it might be that I am missing something but in terms of bounding boxes and metrics the result seems to be identical when running feat/issue-121/extract-vertical-lines vs main
